src/idl/src/keylist.c

```diff
@@ -16,39 +16,171 @@
+static struct key_container *key_container_new (const idl_node_t *node)
 {
-  for (size_t n = 0; n < len; n++)
-    if (strcmp(a[n], b[n]))
-      return false;
-  return true;
+  struct key_container *kc;
+  if ((kc = idl_malloc (sizeof (*kc))) == NULL)
+    return NULL;
+  kc->node = node;
+  return boxed_vector_init_with_failure_action (&kc->key_fields, idl_free, kc);
+}
+
+static void key_container_free (struct key_container *kc) {
+  idl_boxed_vector_fini (&kc->key_fields, key_field_free_wrapper);
+  idl_free (kc);
+}
+static void key_container_free_wrapper (void *vkc) {
+  key_container_free (vkc);
 }
 
-static bool has_conflicting_keys(const struct key_container *key_containers, size_t n_key_containers, const idl_node_t **node)
+static struct key_field *key_container_append (struct key_container *kc, const idl_declarator_t *declarator)
 {
-  assert(node);
-  for (size_t c = 0; c < n_key_containers; c++) {
-    const struct key_container *cntr = &key_containers[c];
+  struct key_field *kf;
+  if ((kf = key_field_new (declarator)) == NULL)
+    return NULL;
+  return boxed_vector_append_with_failure_action (&kc->key_fields, kf, key_field_free_wrapper);
+}
 
+static const struct key_field *key_container_first_c (const struct key_container *kc, struct key_container_iter *it) {
+  return idl_boxed_vector_first_c (&kc->key_fields, &it->x);
+}
+static const struct key_field *key_container_next_c (struct key_container_iter *it) {
+  return idl_boxed_vector_next_c (&it->x);
+}
+static struct key_field *key_container_first (struct key_container *kc, struct key_container_iter *it) {
+  return idl_boxed_vector_first (&kc->key_fields, &it->x);
+}
+static struct key_field *key_container_next (struct key_container_iter *it) {
+  return idl_boxed_vector_next (&it->x);
+}
+
+static struct key_containers *key_containers_new (void)
+{
+  struct key_containers *kcs;
+  if ((kcs = idl_malloc (sizeof (*kcs))) == NULL)
+    return NULL;
+  return boxed_vector_init_with_failure_action (&kcs->key_containers, idl_free, kcs);
+}
+
+static void key_containers_free (struct key_containers *kcs)
+{
+  idl_boxed_vector_fini (&kcs->key_containers, key_container_free_wrapper);
+  idl_free (kcs);
+}
+
+static struct key_container *key_containers_append (struct key_containers *kcs, const idl_node_t *key_type_node)
+{
+  struct key_container *kc;
+  if ((kc = key_container_new (key_type_node)) == NULL)
+    return NULL;
+  return boxed_vector_append_with_failure_action (&kcs->key_containers, kc, key_container_free_wrapper);
+}
+
+static const struct key_container *key_containers_first (const struct key_containers *kcs, struct key_containers_iter *it) {
+  return idl_boxed_vector_first_c (&kcs->key_containers, &it->x);
+}
+static const struct key_container *key_containers_next (struct key_containers_iter *it) {
+  return idl_boxed_vector_next_c (&it->x);
+}
+static struct key_container *key_containers_first_nc (struct key_containers *kcs, struct key_containers_iter *it) {
```

Comment:
Used suffix `_nc` not consistent with `key_container_first_c` above.

Reply:
Agree